dronegen/container_images.go

```diff
@@ -0,0 +1,942 @@
+		Volumes: dockerVolumeRefs(),
+		Environment: map[string]value{
+			"DOCKER_BUILDKIT": {
+				raw: "1",
+			},
+		},
+		Commands: []string{
+			"docker run --privileged --rm tonistiigi/binfmt --install all",
+			fmt.Sprintf("mkdir -pv %q && cd %q", p.WorkingDirectory, p.WorkingDirectory),
+			fmt.Sprintf("mkdir -pv %q", buildxConfigFileDir),
+			fmt.Sprintf("echo '[registry.%q]' > %q", localRegistry, buildxConfigFilePath),
+			fmt.Sprintf("echo '  http = true' >> %q", buildxConfigFilePath),
+			buildxCreateCommand,
+			buildCommand,
+			fmt.Sprintf("docker buildx rm %q", builderName),
+			fmt.Sprintf("rm -rf %q", buildxConfigFileDir),
+		},
+	}
+
+	return step, &buildStepOutput{
+		StepName:   step.Name,
+		BuiltImage: localRegistryImage,
+		Version:    version,
+		Product:    p,
+	}
+}
+
+// The `step` struct doesn't contain enough information to setup
+// dependent steps so we add that via this struct
+type buildStepOutput struct {
+	StepName   string
+	BuiltImage *Image
+	Version    *releaseVersion
+	Product    *Product
+}
+
+type ContainerRepo struct {
+	Name             string
+	IsProductionRepo bool
+	IsImmutable      bool
+	Environment      map[string]value
+	RegistryDomain   string
+	RegistryOrg      string
+	LoginCommands    []string
+	TagBuilder       func(baseTag *ImageTag) *ImageTag // Postprocessor for tags that append CR-specific suffixes
+}
+
+func NewEcrContainerRepo(accessKeyIDSecret, secretAccessKeySecret, domain string, isProduction, isImmutable, guaranteeUnique bool) *ContainerRepo {
+	nameSuffix := "staging"
+	ecrRegion := StagingEcrRegion
+	loginSubcommand := "ecr"
+	if isProduction {
+		nameSuffix = "production"
+		ecrRegion = PublicEcrRegion
+		loginSubcommand = "ecr-public"
+	}
+
+	registryOrg := ProductionRegistryOrg
+	if configureForPRTestingOnly {
+		accessKeyIDSecret = testingSecretPrefix + accessKeyIDSecret
+		secretAccessKeySecret = testingSecretPrefix + secretAccessKeySecret
+		registryOrg = testingECRRegistryOrg
+
+		if !isProduction {
+			domain = testingECRDomain
+			ecrRegion = testingECRRegion
+		}
+	}
+
+	return &ContainerRepo{
+		Name:             fmt.Sprintf("ECR - %s", nameSuffix),
+		IsProductionRepo: isProduction,
+		IsImmutable:      isImmutable,
+		Environment: map[string]value{
+			"AWS_ACCESS_KEY_ID": {
+				fromSecret: accessKeyIDSecret,
+			},
+			"AWS_SECRET_ACCESS_KEY": {
+				fromSecret: secretAccessKeySecret,
+			},
+		},
+		RegistryDomain: domain,
+		RegistryOrg:    registryOrg,
+		LoginCommands: []string{
+			"apk add --no-cache aws-cli",
+			"TIMESTAMP=$(date -d @\"$DRONE_BUILD_CREATED\" '+%Y%m%d%H%M')",
+			fmt.Sprintf("aws %s get-login-password --region=%s | docker login -u=\"AWS\" --password-stdin %s", loginSubcommand, ecrRegion, domain),
+		},
+		TagBuilder: func(tag *ImageTag) *ImageTag {
+			if guaranteeUnique {
+				tag.AppendString("$TIMESTAMP")
+			}
+
+			return tag
+		},
+	}
+}
+
+func NewQuayContainerRepo(dockerUsername, dockerPassword string) *ContainerRepo {
```

Comment:
Out of scope of this PR, but what are our plans for getting rid of quay.io completely? @logand22 maybe?

Reply:
I wasn't sure on our plans to cut out quay.io completely. There are still quite a lot of pulls for it.